KDEpy/mirroring.py

```
@@ -0,0 +1,136 @@
+        """
+        # Round the data to the specified number of decimals.
+        # Important! Otherwise, due to numerical precision, the numbers won't match
+        rounded_data = np.round(mirrored_data, decimals=decimals)
+
+        unique_data, indices = np.unique(
+            rounded_data, axis=0, return_inverse=True
+        )  # Find unique rows and their indices
+        summed_values = np.zeros(unique_data.shape[0])  # Preallocate the summed values array
+        np.add.at(summed_values, indices, updated_values)  # Use np.add.at to sum the values for each unique row
+
+        return unique_data, summed_values
+
+    if pdf_values is None:
+        pdf_values = (
+            np.ones(data.shape[0]) / data.shape[0]
+        )  # If no PDF values are provided, assume uniform distribution.
+
+    assert pdf_values.shape[0] == data.shape[0], "PDF values must match the data size."
+
+    mirrored_data = data.copy()
+    updated_values = pdf_values.copy()
+
+    # Each boundary is 'folding' the data, and creating a copy, and adding the PDF values.
+    for dim, boundary in enumerate(boundaries):
+        if boundary is not None:
+            lower, upper = boundary
+            if lower is not None:
+                try:
+                    closest_lower = np.max(
+                        data[data[:, dim] <= lower, dim]
+                    )  # Mirror using the closest data point as pivot
+                    lower_mirror = 2 * closest_lower - data[:, dim]
+                    mirrored_points = np.column_stack(
+                        [lower_mirror if i == dim else data[:, i] for i in range(data.shape[1])]
+                    )
+                    mirrored_data = np.vstack([mirrored_data, mirrored_points])
+                    updated_values = np.concatenate([updated_values, pdf_values])
+
+                    mirrored_data, updated_values = sum_together_np(mirrored_data, updated_values)
+                except Exception as e:
```

I dislike these blanket try/except statements. How exactly can the algorithm fail? If it fails, it is really good enough to just keep going? Presumably something went wrong, and then I think it's better to just raise the exception. Returning bad results does not do the caller any favors.

---

You're right, should be better now.